common/wallet_tx.c

```diff
@@ -39,6 +48,59 @@ struct command_result *param_wtx(struct command *cmd,
         return NULL;
 }
 
+struct command_result *param_wtx_utxos(struct command *cmd,
+				 const char *name,
+				 const char *buffer,
+				 const jsmntok_t *tok,
+				 struct wallet_tx *wtx)
+{
+	size_t i;
+	const jsmntok_t *curr;
+	struct bitcoin_txid **txids;
+	unsigned int **outnums;
+	char **outpoint = tal_arr(tmpctx, char*, 0);
+
+	txids = tal_arr(cmd, struct bitcoin_txid*, 0);
+	outnums = tal_arr(cmd, unsigned int*, 0);
+	json_for_each_arr(i, curr, tok) {
+		outpoint = tal_strsplit(cmd, json_strdup(cmd, buffer, curr), ":",
+				STR_NO_EMPTY);
+		if (tal_count(outpoint) < 3 || strlen(outpoint[0]) != 64)
+			break;
+
+		struct bitcoin_txid *txid = tal(tmpctx, struct bitcoin_txid);
+		unsigned int *outnum = tal(tmpctx, unsigned int);
+		if (!bitcoin_txid_from_hex(outpoint[0], strlen(outpoint[0]), txid)
+				|| opt_set_uintval(outpoint[1], outnum))
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				"Could not decode the outpoint with txid \"%s\" and "
+				"output_index \"%s\". The utxos should be"
+				" specified as 'txid:output_index'.", outpoint[0], outpoint[1]);
+		tal_arr_expand(&txids, txid);
+		tal_arr_expand(&outnums, outnum);
+	}
+
+	if (!tal_count(txids) || !tal_count(outnums))
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+			"Please specify an amount in satoshis, 'all' or an array"
+			" of 'txid:output_index'.");
+
+	wtx->utxos = wallet_select_specific(cmd, cmd->ld->wallet, txids, outnums);
+	tal_free(txids);
```

Comment:
if the `command_fail`s at any point previous to this, do we need to call `tal_free` for each of those branches as well?

Reply:
No because tal allocated variables passed to this function are off `cmd`, which is desallocated on `comand_error` and then so they are